chips/nrf52/src/i2c.rs

```diff
@@ -0,0 +1,446 @@
+//! Implementation of I2C for nRF52 using EasyDMA.
+//!
+//! This module supports nRF52's two I2C master (`TWIM`) peripherals,
+//! but not I2C slave (`TWIS`).
+//!
+//! - Author: Jay Kickliter
+//! - Author: Andrew Thompson
+//! - Date: Nov 4, 2017
+
+use core::cell::Cell;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil;
+use nrf5x::nvic;
+use nrf5x::peripheral_interrupts::NvicIdx;
+use nrf5x::pinmux::Pinmux;
+
+/// An I2C master device.
+///
+/// A `TWIM` instance wraps a `registers::TWIM` together with
+/// addition data necessary to implement an asynchronous interface.
+pub struct TWIM {
+    registers: *const registers::TWIM,
+    nvic_idx: NvicIdx,
+    client: Cell<Option<&'static hil::i2c::I2CHwMasterClient>>,
+    busy: Cell<bool>,
+    buf: TakeCell<'static, [u8]>,
+    enabled: Cell<bool>,
+}
+
+impl TWIM {
+    const fn new(instance: usize) -> TWIM {
+        TWIM {
+            registers: registers::INSTANCES[instance].0,
+            nvic_idx: registers::INSTANCES[instance].1,
+            client: Cell::new(None),
+            busy: Cell::new(false),
+            buf: TakeCell::empty(),
+            enabled: Cell::new(false),
```

Comment:
Is this field necessary? Can't we always get the state by reading the `ENABLE` register?

Reply:
Possibly, but due to the fact that SPI & I2C share registers, I pulled it out.

Reply:
Right, but can't `is_enabled` just be implemented by reading the register? Or is exactly a problem because that's actually a shared register between TWIM/TWIS/SPI?

Reply:
They have different values depending on which peripheral is being used. I believe we could read the register here.

Reply:
I'm fine with keeping these if that's simpler somehow. It's not a lot of extra space or anything.

It _kind of_ sounds like maybe SPIM+TWIM+TWIS should really end up being merged (since, oddly, they live in the same registers and we just didn't realize that) at which point it may make sense to revisit this, but I don't think it's critical to figure that out for now.

Maybe just a comment somewhere hinting that's what's going on (that they share registers).